include/nbl/builtin/hlsl/subgroup/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,490 @@
+
+[[vk::ext_instruction(356)]]
+int spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, int value);
+[[vk::ext_instruction(357)]]
+uint spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, uint value);
+[[vk::ext_instruction(358)]]
+float spirv_subgroupPrefixMax(uint scope, [[vk::ext_literal]] uint operation, float value);
+
+template<>
+struct inclusive_scan<int, binops::max<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::max<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct inclusive_scan<uint, binops::max<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMax(3, 1, x);
+    }
+};
+
+template<>
+struct exclusive_scan<int, binops::max<int> >
+{
+    int operator()(const int x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::max<uint> >
+{
+    uint operator()(const uint x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+template<>
+struct exclusive_scan<uint, binops::max<float> >
+{
+    float operator()(const float x)
+    {
+        return spirv_subgroupPrefixMax(3, 2, x);
+    }
+};
+
+}
+#else
+namespace portability
+{
+template<typename T, class Binop, class ScratchAccessor, bool initializeScratch>
+struct inclusive_scan
+{
+    static inclusive_scan<T, Binop, ScratchAccessor, initializeScratch> create()
+    {
+		inclusive_scan<T, Binop, ScratchAccessor, initializeScratch> retval;
+		retval.offsetsAndMasks = ScratchOffsetsAndMasks::WithSubgroupOpDefaults();
+		return retval;
+    }
+
+    T operator()(T value)
+    {
+		Binop op;
+
+		if (initializeScratch)
+		{
+			Barrier();
+			MemoryBarrierShared();
+			scratchInitialize<ScratchAccessor, T>(value, op.identity(), _NBL_HLSL_WORKGROUP_SIZE_-1);
+		}
+		Barrier();
+		MemoryBarrierShared();
+		
+		// Stone-Kogge adder
+		// (devsh): it seems that lanes below <HalfSubgroupSize/step are doing useless work,
+		// but they're SIMD and adding an `if`/conditional execution is more expensive
+	#if 1 // Use shuffling by default (either native or portability implementation)
+		uint toAdd = ShuffleUp<T, ScratchAccessor>(value, 1u); // all invocations must execute the shuffle, even if we don't apply the op() to all of them
+		if(offsetsAndMasks.subgroupInvocation >= 1u) {
+			// the first invocation (index 0) in the subgroup doesn't have anything in its left
+			value = op(value, toAdd);
+		}
+	#else
+		value = op(value, scratch.main.get(offsetsAndMasks.scanStoreOffset-1u));
+	#endif
+		[[unroll]] // REVIEW: I copied this from the example implementation on the github issue but if I'm not mistaken unroll doesn't work since halfSubgroupSize is not constexpr 
+		for (uint step=2u; step <= offsetsAndMasks.halfSubgroupSize; step <<= 1u)
```

Comment:
unroll takes a hint, you can tell if how many times to unroll actually even if the condition is not a cosntexpr